### PR TITLE
feat(translator): early request header modifier

### LINKED ISSUE #3978: Support early header mutation
*Description*:
Envoy supports early header mutations, that occur prior to any routing and built-in envoy header processing. These mutations can be useful for preprocessing/preservation/injection of headers before envoy intervenes. 

Some use cases described here:
- https://github.com/istio/istio/issues/49978 
- https://github.com/envoyproxy/envoy/issues/23786 
- https://github.com/envoyproxy/envoy/issues/34589

**Proposal:**
Extend ClientTrafficPolicy's `headers` container to contain early request/response header modification, reusing the structs from similar GW-API HTTP route filters. 

**Note:** Custom Envoy Gateway HTTP route filter is not applicable here, since early mutations are configured on HCM level, instead of route level. 

*Relevant Links*:
- https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/http/early_header_mutation
- https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
- https://gateway-api.sigs.k8s.io/guides/http-header-modifier/?h=requestheadermodifier#http-request-header-modifier

### PR DESCRIPTION
**Pull request by guydc**
**What type of PR is this?**

**What this PR does / why we need it**:
Adds support for early request header mutation in client traffic policy
**Which issue(s) this PR fixes**:
Fixes #3978

### CODE REVIEW

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 1]
```diff
@@ -869,3 +883,130 @@ func buildConnection(connection *egv1a1.ClientConnection) (*ir.ClientConnection,
...
+		}
+		for _, addHeader := range headersToAdd {
+			emptyFilterConfig = false
+			if addHeader.Name == "" {
+				errs = errors.Join(errs, fmt.Errorf("EarlyRequestHeaderModifier cannot add a header with an empty name"))
+				// try to process the rest of the headers and produce a valid config.
+				continue
+			}
+			// Per Gateway API specification on HTTPHeaderName, : and / are invalid characters in header names
+			if strings.Contains(string(addHeader.Name), "/") || strings.Contains(string(addHeader.Name), ":") {
```
Comment (liorokman) [inline internal/gatewayapi/clienttrafficpolicy.go:911]: Maybe use [strings.ContainsAny](https://pkg.go.dev/strings#ContainsAny) instead?

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 2]
```diff
@@ -869,3 +883,130 @@ func buildConnection(connection *egv1a1.ClientConnection) (*ir.ClientConnection,
...
+			emptyFilterConfig = false
+		}
+		for _, setHeader := range headersToSet {
+
+			if setHeader.Name == "" {
+				errs = errors.Join(errs, fmt.Errorf("EarlyRequestHeaderModifier cannot set a header with an empty name"))
+				continue
+			}
+			// Per Gateway API specification on HTTPHeaderName, : and / are invalid characters in header names
+			if strings.Contains(string(setHeader.Name), "/") || strings.Contains(string(setHeader.Name), ":") {
```
Comment (liorokman) [inline internal/gatewayapi/clienttrafficpolicy.go:951]: `ContainsAny` instead of multiple `Contains` here as well.

**Review by liorokman — APPROVED** [review]

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 3]
```diff
@@ -134,6 +135,12 @@ type HeaderSettings struct {
 	//
 	// +optional
 	PreserveXRequestID *bool `json:"preserveXRequestID,omitempty"`
+
+	// EarlyRequestHeaderModifier defines settings for early request header modification, before envoy performs
+	// routing, tracing and built-in header manipulation.
+	//
+	// +optional
+	EarlyRequestHeaderModifier *gwapiv1.HTTPHeaderFilter `json:"earlyRequestHeaderModifier,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/clienttrafficpolicy_types.go:143]: thoughts on `earlyRequestHeaders` ?

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**